core/chain/src/mempool.rs

```diff
@@ -0,0 +1,112 @@
+use std::collections::HashSet;
+use std::sync::Arc;
+use parking_lot::RwLock;
+use tokio::sync::mpsc::{Sender, Receiver};
+use futures::{Future, Stream};
+use log::error;
+use transaction::SignedTransaction;
+use crate::types::{ReceiptBlock, ChainPayload, SignedShardBlock};
+
+/// mempool that stores transactions and receipts for a chain
+#[derive(Default)]
+struct Pool {
+    transactions: HashSet<SignedTransaction>,
+    receipts: HashSet<ReceiptBlock>,
+}
+
+impl Pool {
```

Comment:
So pool must do next things:
 - On insertion of tx -> be able fast query if the new transaction already in the accepted blocks (each pool is attached this way to a chain, and can use TransactionAddress there).
 - On new block prune the tx (implemented)
 - Produce payload (implemented)
 - Be able to tell which tx are missing given some info (bloomfilter?) and query for them
 - Be able to respond to such queries with tx.
Last two things are best looked at eth parity - 

Additional stuff:
 - temporary ban transactions that are invalid - e.g. wrong signatures, missing account etc
 - order transacitons on importing based on account_id / nonce to actually execute them in the order they meant to be. Grouping by account_id can also help us later to execute them in parallel.
  - limit size of the pool to not eat all RAM.


Generally, https://github.com/paritytech/substrate/blob/master/core/transaction-pool/graph and https://github.com/paritytech/parity-ethereum/blob/d89b8d904ff28382d3692290a936706304627255/miner/src/pool/ is a good thing to look at for inspiration of functionality.

Reply:
Posted this comments into #576

Reply:
Don't need to do all of it one PR, just wanted to make sure we have this list.